python/taichi/lang/field.py

```diff
@@ -0,0 +1,459 @@
+    def __str__(self):
+        if impl.inside_kernel():
+            return self.__repr__()  # make pybind11 happy, see Matrix.__str__
+        else:
+            return str(self.to_numpy())
+
+    def pad_key(self, key):
+        if key is None:
+            key = ()
+        if not isinstance(key, (tuple, list)):
+            key = (key, )
+        assert len(key) == len(self.shape)
+        return key + ((0, ) * (_ti_core.get_max_num_indices() - len(key)))
+
+    def initialize_host_accessors(self):
+        if self.host_accessors:
+            return
+        impl.get_runtime().materialize()
+        self.host_accessors = [
+            SNodeHostAccessor(e.ptr.snode()) for e in self.vars
+        ]
+
+
+class ScalarField(Field):
+    """Taichi scalar field with SNode implementation.
+
+    Args:
+        var (Expr): Field member.
+    """
+    def __init__(self, var):
+        super().__init__([var])
+
+    @python_scope
+    def fill(self, val):
+        from taichi.lang.meta import fill_tensor
+        fill_tensor(self, val)
+
+    @python_scope
+    def to_numpy(self, dtype=None):
+        if dtype is None:
+            dtype = to_numpy_type(self.dtype)
+        import numpy as np
+        arr = np.zeros(shape=self.shape, dtype=dtype)
+        from taichi.lang.meta import tensor_to_ext_arr
+        tensor_to_ext_arr(self, arr)
+        ti.sync()
+        return arr
+
+    @python_scope
+    def to_torch(self, device=None):
+        import torch
+        arr = torch.zeros(size=self.shape,
+                          dtype=to_pytorch_type(self.dtype),
+                          device=device)
+        from taichi.lang.meta import tensor_to_ext_arr
+        tensor_to_ext_arr(self, arr)
+        ti.sync()
+        return arr
+
+    @python_scope
+    def from_numpy(self, arr):
+        assert len(self.shape) == len(arr.shape)
+        for i in range(len(self.shape)):
+            assert self.shape[i] == arr.shape[i]
+        if hasattr(arr, 'contiguous'):
+            arr = arr.contiguous()
+        from taichi.lang.meta import ext_arr_to_tensor
+        ext_arr_to_tensor(arr, self)
+        ti.sync()
+
+    @python_scope
+    def __setitem__(self, key, value):
+        self.initialize_host_accessors()
+        self.host_accessors[0].setter(value, *self.pad_key(key))
+
+    @python_scope
+    def __getitem__(self, key):
+        self.initialize_host_accessors()
+        return self.host_accessors[0].getter(*self.pad_key(key))
+
+    def __repr__(self):
+        # make interactive shell happy, prevent materialization
+        return '<ti.field>'
+
+
+class MatrixField(Field):
+    """Taichi matrix field with SNode implementation.
+
+    Args:
+        vars (Expr): Field members.
+        n (Int): Number of rows.
+        m (Int): Number of columns.
+    """
+    def __init__(self, vars, n, m):
+        assert len(vars) == n * m
+        super().__init__(vars)
+        self.n = n
+        self.m = m
+
+    def get_scalar_field(self, *indices):
```

Comment:
Hmm, wehave to keep `__call__()` for a while to help ppl migrate, if necessary. Consider adding `__call__()` and mark it as deprecated?

Reply:
What about the test cases where I replaced the `__call__()`?

Reply:
Fine to update the test cases